### PR TITLE
[Improvement] Update Candidate with multi-dim search constraints.

### PR DESCRIPTION
**Pull request by aptsunny**
Part of [Autoformer algo](https://github.com/open-mmlab/mmrazor/pull/288)

## Motivation
Continued improvements of EvolutionSearchLoop throught multi-resources constraints.

## Use cases (Optional)

Modify the search configs file as follow.

```python
flops_range=(0., 330.) ->  constraints_range=dict(flops=(0., 330.)),
```

```python
train_cfg = dict(
    _delete_=True,
    type='mmrazor.EvolutionSearchLoop',
    dataloader=_base_.val_dataloader,
    evaluator=_base_.val_evaluator,
    max_epochs=20,
    num_candidates=20,
    top_k=10,
    num_mutation=5,
    num_crossover=5,
    mutate_prob=0.2,
    constraints_range=dict(params=(0, 55), flops=(0., 330.)),
    dump_derived_mutable=True,
    score_key='accuracy/top1')
```
## Checklist

**After PR**:

- [ ] The search configuration file for the search algorithm needs to be changed. Please fix it

### CODE REVIEW

File: mmrazor/structures/subnet/candidate.py [hunk 1]
```diff
@@ -88,4 +137,43 @@ def extend(self, other: Any) -> None:
 
     def set_score(self, i: int, score: float) -> None:
         """Set score to the specified subnet by index."""
-        self.data[i] = (self.data[i][0], float(score))
+        for _, value in self.data[i].items():
+            value['score'] = score
+
+    def set_resources(self,
```
Comment (sunnyxiaohu) [inline mmrazor/structures/subnet/candidate.py:143]: set_resources -> set_resource

File: mmrazor/structures/subnet/candidate.py [hunk 2]
```diff
@@ -88,4 +137,43 @@ def extend(self, other: Any) -> None:
 
     def set_score(self, i: int, score: float) -> None:
```
Comment (sunnyxiaohu) [inline mmrazor/structures/subnet/candidate.py:138]: `set_score` could call `set_resource` by using key_indicator='score'

File: mmrazor/engine/runner/evolution_search_loop.py [hunk 3]
```diff
@@ -283,28 +322,34 @@ def _save_searcher_ckpt(self) -> None:
...
             subnet=random_subnet,
             estimator=self.estimator,
-            flops_range=self.flops_range)
+            constraints_range=self.constraints_range)
+
+        return is_pass, results
+
 
-        return is_pass
+if __name__ == '__main__':
```
Comment (sunnyxiaohu) [inline mmrazor/engine/runner/evolution_search_loop.py:353]: clean

File: mmrazor/engine/runner/utils/check.py [hunk 4]
```diff
@@ -15,18 +13,19 @@
...
+    model,
+    subnet: SupportRandomSubnet,
+    estimator: ResourceEstimator,
+    constraints_range: Dict[str, Any] = dict(flops=(0, 330))
+) -> Tuple[bool, Dict]:
+    """Check whether is beyond resources constraints.
 
     Returns:
         bool: The result of checking.
     """
```
Comment (sunnyxiaohu) [inline mmrazor/engine/runner/utils/check.py:26]: update docstring: Returns

File: tests/test_runners/test_utils/test_check.py [hunk 5]
```diff
@@ -12,29 +12,38 @@
...
     mock_estimator.estimate.return_value = fake_results
-    result = check_subnet_flops(mock_model, fake_subnet, mock_estimator,
-                                flops_range)
-    assert result is False
+    is_pass, _ = check_subnet_resources(mock_model, fake_subnet,
+                                        mock_estimator, constraints_range)
+    assert is_pass is False
+
+
+if __name__ == '__main__':
```
Comment (sunnyxiaohu) [inline tests/test_runners/test_utils/test_check.py:47]: clear

**Review by sunnyxiaohu — APPROVED** [review]

File: tests/test_runners/test_evolution_search_loop.py [hunk 6]
```diff
@@ -103,64 +103,70 @@ def test_init(self):
...
+    @patch('mmrazor.models.task_modules.estimators.resource_estimator.'
+           'get_model_flops_params')
+    def test_run_epoch(self, flops_params, mock_export_fix_subnet,
+                       load_status):
         # test_run_epoch: distributed == False
         loop_cfg = copy.deepcopy(self.train_cfg)
         loop_cfg.runner = self.runner
         loop_cfg.dataloader = self.dataloader
         loop_cfg.evaluator = self.evaluator
+        loop_cfg.runner.epoch = 1
```
Comment (gaoyang07) [inline tests/test_runners/test_evolution_search_loop.py:125]: check the usage of `runner.epoch`(`loop._epoch`) in UT

File: mmrazor/engine/runner/subnet_sampler_loop.py [hunk 7]
```diff
@@ -177,10 +175,7 @@ def __init__(self,
 
         self.candidates = Candidates()
         self.top_k_candidates = Candidates()
-        if resource_estimator_cfg is None:
-            self.estimator = ResourceEstimator()
-        else:
-            self.estimator = ResourceEstimator(**resource_estimator_cfg)
+        self.estimator = ResourceEstimator(**resource_estimator_cfg)
```
Comment (gaoyang07) [inline mmrazor/engine/runner/subnet_sampler_loop.py:178]: use `TASK_UTILS` to build estimator

**Review by gaoyang07 — APPROVED** [review]

**Pull request merged**